### ISSUE TITLE
LightFM always produces same precision@6 result metric

### ISSUE THREAD

**Issue by scottee** [lyst/lightfm#148]
I'm running parameter optimization on a LightFM implicit factorization model using "warp" loss.  When I run np.mean(precision_at_k(...)) on the test data, it is virtually always the same result, 0.316850870848, out to 12 digits.  Is this expected for some reason?  If not, any ideas how to figure out what might be wrong?

Here is sample output showing the param combinations and the output:

```
2017-01-20 21:10:02,363 INFO __main__: Starting training iteration 1, params:
{
    "alpha": 0.0041233076328981919,
    "epochs": 45,
    "learning_rate": 0.50174314850490254,
    "no_components": 184
}
2017-01-20 21:10:02,363 INFO lightfm_uma: Training model...
2017-01-20 21:25:20,518 INFO lightfm_uma: Finished precision@6 = 0.316850870848
2017-01-20 21:25:21,453 INFO __main__: Starting training iteration 2, params:
{
    "alpha": 0.0064873564172718886,
    "epochs": 63,
    "learning_rate": 0.50406151543722921,
    "no_components": 180
}
2017-01-20 21:25:21,453 INFO lightfm_uma: Training model...
2017-01-20 21:44:36,565 INFO lightfm_uma: Finished precision@6 = 0.316850870848
2017-01-20 21:44:37,495 INFO __main__: Starting training iteration 3, params:
{
    "alpha": 0.020473212242717205,
    "epochs": 62,
    "learning_rate": 0.74135691825946459,
    "no_components": 156
}
2017-01-20 21:44:37,496 INFO lightfm_uma: Training model...
2017-01-20 22:00:45,661 INFO lightfm_uma: Finished precision@6 = 0.316850870848
```

**Comment by maciejkula** [comment]
My suspicion is that in this case your learning rate is too large and the
model diverges. Can you check the actual predicted scores?

**Comment by maciejkula** [comment]
Actually, I think this is a better explanation: you have high regularization parameters, and everything is regularized away to zero. All the predictions are zero too, and the resulting precision values are nonsensical.

This would change that to be less misleading: https://github.com/lyst/lightfm/pull/149

**Comment by scottee** [comment]
I'll look at what the predictions say and report back.  And for clarification, I'm using the adagrad learning_schedule.

BTW, what parameter equates to regularization, alpha or learning_rate?

In the mean time, do you have recommendations on ranges for alphas and learning_rate?  Below is the space I'm using for parameter optimization.  Is this reasonable?

    space = [
        (10**-6, 10**-1, 'log-uniform'),   # alpha (user_alpha and item_alpha)
        (20, 200),                                   # epochs
        (10**-4, 1.0, 'log-uniform'),      # learning_rate
        (20, 200),                                  # no_components
    ]

Below are the lowest alpha and learning_rate combos where the magic precision@6 number happens.  These don't seem very high, but I've never trained with a Warp/BPR type approach before.

 {"alpha": 0.07356951575008387, "epochs": 1, "learning_rate": 0.054121552105164714, "no_components": 48, "precision@6": 0.316850870848},
{"alpha": 0.00684464334830365, "epochs": 187, "learning_rate": 0.005631464052454842, "no_components": 154, "precision@6": 0.316850870848},
{"alpha": 0.001769098654581489, "epochs": 70, "learning_rate": 0.06384023381248471, "no_components": 87, "precision@6": 0.316850870848},
{'alpha': 0.0039032887003676189, 'epochs': 84, 'learning_rate': 0.065409138131584846, 'no_components': 157, 'precision@6': 0.316850870848},

**Comment by scottee** [comment]
Attached is a python session that looks at the predictions when the weird precision@6 does and doesn't happen.  It show several things:

- When the precision@6 goes weird, all predictions are either NaN or 0.  How can predictions being NaN be correct?
- When the precision@6 goes weird, the precision@6 values are often above 1.0.  How can this be right?
- When predictions are all NaN or 0, how can there be any precision@6 value but 0?
- When precsion@6 is sane, all predictions are > 0 and <= 1.0.

So it really seems like several problems, possibly with a common root cause (predictions going to NaN?).  Thoughts?
[lightfm_problem.txt.zip](https://github.com/lyst/lightfm/files/721593/lightfm_problem.txt.zip)

**Comment by maciejkula** [comment]
This is an artifact of how rank is computed when there are ties. When everythig is tied (the 0.0 zero case), all positive examples have rank one. The PR I linked to above (https://github.com/lyst/lightfm/pull/149) fixes this and would make your precision 0.0. Which I guess isn't the right solution either, but at least it indicates that your model has a problem.

**Comment by scottee** [comment]
Where do the NaN predictions come from, though?

**Comment by maciejkula** [comment]
All SGD algorithms are sensitive to hyperparameters. NaNs are most often the result of excessive learning rates.

**Comment by scottee** [comment]
Yes, agreed.  But the learning rates are not excessive in all cases where this happens (at least in my experience).  In your experience, what should the range on learning rate be?

**Comment by maciejkula** [comment]
Lower than 0.1 is probably a good idea. However, this also depends on data, and on the scale of the features you are using.

**Comment by scottee** [comment]
I'm just doing simple IMF, no user or item features.  Matrix cells are either 1 or 0.

Can you point me to any place in the code to start debugging?  Since we have a case where it happens on epoch 0, we have a chance of seeing the NaNs happen.

I'll also try resetting the learning_rate range and optimizing again...

And BTW, thanks for all the help!

**Comment by maciejkula** [comment]
I would do the following:

1. Start with a known faulty configuration.
2. Start eliminating other confounding factors, like number of epochs or regularization.
3. Stop when the problem disappears.

That should give us the culprit. To help, I made the model throw an exception when it diverges: https://github.com/lyst/lightfm/pull/151

**Comment by scottee** [comment]
Just FYI... Running optimizations constrained to 0.1 max learning_rate.  There have been 4 weird precision@6 numbers over 115 optimization combinations.   Seems to make a case for the model divergence theory.

Still debugging where the NaN predictions are coming from.

**Comment by scottee** [comment]
Several new points:

- The dev version doesn't diverge on several of the cases where 1.11 does diverge.  So the dev version has some improvements that avoid some of the divergences.
- I found one combination where the model still diverges.  It diverges on item_embeddings going +/-inf.  I don't see an obvious cause of that, but you guys are the experts.

Here's the param combo that still causes divergence:

  {"alpha": 0.07356951575008387, "epochs": 1, "learning_rate": 0.054121552105164714, "no_components": 48}

These all look reasonable to me.  Am I missing something?  And this combo continues to diverge until I reduce the learning rate to 0.0005.  That's an extremely small rate, isn't it?

And here's the session showing the +/- inf:

(Pdb) l
320  	        for parameter in (self.item_embeddings,
321  	                          self.item_biases,
322  	                          self.user_embeddings,
323  	                          self.user_biases):
324  	            if not np.all(np.isfinite(parameter)):
325  ->	                raise ValueError("Not all estimated parameters are finite, "
326  	                                 "your model may have diverged. Try decreasing "
327  	                                 "the learning rate.")
328
329  	    def fit(self, interactions,
330  	            user_features=None, item_features=None,
(Pdb) parameter.shape
(2001, 48)
(Pdb) self.item_embeddings.shape
(2001, 48)
(Pdb) np.isnan(parameter).sum()
0
(Pdb) (parameter==np.inf).sum()
46
(Pdb) (parameter==-np.inf).sum()
72
(Pdb) (parameter==0).sum()
95930

The +/- inf happens in the first call to fit_warp().  I can't seem to debug this further since the code is all .pyx -- unless there's magic to do that.  (I've never debugged at the pyx level.)

**Comment by maciejkula** [comment]
Thank you for all the work digging into this!

Some assorted thoughts:
1. Whether the fitting diverges or not can be non-deterministic. You may
want to fix the random seed and disable parallelism to get repeatable runs.
2. You can insert any Python code into the Cython files. You will have to
disable OpenMP to do that, though. Some print statements may show you the
evolution of parameter values as you're fitting the model.

In general I agree that your parameter values look sensible and the
algorithm should work with these values.

**Comment by scottee** [comment]
I already fix the random seed, so I was hoping it is deterministic.  Does parallelism introduce non-determinism also?

 I'll try the good, old debugging by print and see what I find...

**Comment by maciejkula** [comment]
Yes. All the threads write to a shared memory buffer, so results will very
from run to run if you have more than one thread.

**Comment by scottee** [comment]
I may have found an issue, or at least a symptom.  Please tell me if I'm chasing a red herring here.

In file _lifhtfm_fast.pyx.template around line 440-ish there is the regularization on feature+component:

            features[feature, component] *= (1.0 + alpha * local_learning_rate)

Since during an epoch any feature+component cell can be visited many times, this regularization compounds on itself, instead of adding a little noise.  (For example, one cell was regularized 19K times in my log file.) I've not implemented SGD algorithms before, but it seems fishy that the code regularizes a single cell multiple times during an epoch, and especially with multiplication.   Below are some print statements that illustrate the problem.  These were printed during one epoch:

WARN: update_features() reg on features[989, 0] is BIG 1001730.187500, old val 997782.437500, alpha 0.073570, l_rate 0.053779
WARN: update_features() reg on features[989, 0] is BIG 1005693.562500, old val 1001730.187500, alpha 0.073570, l_rate 0.053779
WARN: update_features() reg on features[989, 0] is BIG 1009672.625000, old val 1005693.562500, alpha 0.073570, l_rate 0.053779
... 18852 lines deleted with updates on [989, 0]

Here's another example:

WARN: update_features() reg on features[989, 12] is BIG -inf, old val -339421865411093937693968221348707696640.000000, alpha 0.073570, l_rate 0.053749

Now interestingly, [989, 12] is only updated once via regularization during the epoch, so something else is making the cell value big and regularization is pushing it to inf.  Still need to dig into that.  That might be another problem or the real root cause.

**Comment by scottee** [comment]
Oops.  What I said about [989,12] isn't true.  I had a bug in the check where it didn't check for large negative values, so it didn't print all the times value was neg. large.  So it looks like the regularization is compounding to +/- inf.

**Comment by scottee** [comment]
Another point about the regularization...  I'm seeing lots of cases where the gradient has gone to 0, but then regularization compounds on a value.  That doesn't seem right either.

**Comment by maciejkula** [comment]
I think you're onto the disadvantages of the way I implemented lazy regularization here. It inflates the values of certain features, then divides them by a global regularization value. With high regularization rates this can run away to infinity. 

In other projects I have found a better way of doing this which should be more stable numerically, but requires a lot more memory. I'll need to think about that.

In the meantime, can you try reducing the `alpha` values and see if the problem goes away? (Plus, there is a slight bug in the implementation as well. I'll fix that.)

Again, thanks a lot for looking into this!

**Comment by scottee** [comment]
I already played with alpha.  If I divide alpha by 100 (0.0007) it works.  If I divide learning_rate by 100 (0.0005) it works.  If I divide each by 10 it still diverges.

**Comment by scottee** [comment]
And BTW, I have a non-proprietary set of data and code to reproduce the problem.  Unfortunately the training data is 46MB compressed.  If we can figure how to pass it along I can share it.

**Comment by maciejkula** [comment]
That would be great, I could then go away and hammer at it. A Dropbox link
maybe? If you'd like to send it privately you can contact me via my Twitter
account @Maciej_Kula or the LightFM gitter channel.

**Comment by maciejkula** [comment]
I added a tentative fix for the runaway growth of regularization values in https://github.com/lyst/lightfm/pull/152

This forces a full regularization cycle mid-epoch if necessary, and I _think_ this will resolve your problem.

**Comment by scottee** [comment]
It's a good news, bad news thing:

- Good: The #152 PR does seem to have fixed the divergence problem.
- Bad: It also seems to have crippled performance.
- Good: Crippling performance seems to happen mostly when model starts to diverge, or gradient goes to 0.  Small performance hit w/o openmp.

With openmp and 16 threads, the divergent param combo is barely faster than no openmp and 1 thread.  Here's 16 threads:

[hadoop@ip-10-13-181-155 ~]$ ./lightfm_optimize_prob.py
2017-01-25 16:30:17,882 INFO __main__: Training model...
Epoch 0
2017-01-25 16:36:27,548 INFO __main__: Finished training model

Takes 6min, 10sec.  Here's 1 thread, no openmp, but on my laptop, not AWS:

$ ./lightfm_optimize_prob.py
/Users/escott/projects/lightfm/lightfm/_lightfm_fast.py:9: UserWarning: LightFM was compiled without OpenMP support. Only a single thread will be used.
  warnings.warn('LightFM was compiled without OpenMP support. '
2017-01-25 10:17:55,189 INFO __main__: Training model...
Epoch 0
2017-01-25 10:24:02,712 INFO __main__: Finished training model

Takes 6min 7sec.  Here's 1 thread, with openmp on AWS:

[hadoop@ip-10-13-181-155 ~]$ ./lightfm_optimize_prob.py  -threads 1
2017-01-25 16:45:17,436 INFO __main__: Training model...
Epoch 0
2017-01-25 16:52:12,090 INFO __main__: Finished training model

Takes 6min, 55sec.

Now let's look at a non-diverging, gradient doesn't go to zero, param combo.  One thread, no openmp:

$ ./lightfm_optimize_prob.py
/Users/escott/projects/lightfm/lightfm/_lightfm_fast.py:9: UserWarning: LightFM was compiled without OpenMP support. Only a single thread will be used.
  warnings.warn('LightFM was compiled without OpenMP support. '
2017-01-25 10:50:20,861 INFO __main__: Training model...
...
Epoch 19
2017-01-25 11:01:26,286 INFO __main__: Finished training model

Takes 11min 6sec.  This used to take 10min, 30sec before #152.

Now 16 threads and openmp on AWS:

[hadoop@ip-10-13-181-155 ~]$ ./lightfm_optimize_prob.py
2017-01-25 16:59:48,265 INFO __main__: Training model...
...
Epoch 19
2017-01-25 17:03:57,042 INFO __main__: Finished training model

Takes 4min 9sec.  This isn't significantly slower than it used to be, that I can tell.

**Comment by maciejkula** [comment]
This is expected - at high regularisation rates full regularisation passes
are triggered repeatedly during an epoch, holding up all the threads until
the work is finished. I can probably optimize this a little more.

The result of this change should be that:
1. There should not be any performance implications with low regularisation
rates.
2. When regularisation rates are large, the algorithm will be slow but
numerically stable (as opposed to fast but producing nonsense).

On 25 Jan 2017 09:43, "Eric Scott" <notifications@github.com> wrote:

It's a good news, bad news thing:

   - Good: The #152 <https://github.com/lyst/lightfm/pull/152> PR does seem
   to have fixed the divergence problem.
   - Bad: It also seems to have crippled performance.
   - Good: Crippling performance seems to happen mostly when model starts
   to diverge, or gradient goes to 0. Small performance hit w/o openmp.

With openmp and 16 threads, the divergent param combo is barely faster than
no openmp and 1 thread. Here's 16 threads:

[hadoop@ip-10-13-181-155 ~]$ ./lightfm_optimize_prob.py
2017-01-25 16:30:17,882 INFO *main*: Training model...
Epoch 0
2017-01-25 16:36:27,548 INFO *main*: Finished training model

Takes 6min, 10sec. Here's 1 thread, no openmp, but on my laptop, not AWS:

$ ./lightfm_optimize_prob.py
/Users/escott/projects/lightfm/lightfm/_lightfm_fast.py:9: UserWarning:
LightFM was compiled without OpenMP support. Only a single thread will be
used.
warnings.warn('LightFM was compiled without OpenMP support. '
2017-01-25 10:17:55,189 INFO *main*: Training model...
Epoch 0
2017-01-25 10:24:02,712 INFO *main*: Finished training model

Takes 6min 7sec. Here's 1 thread, with openmp on AWS:

[hadoop@ip-10-13-181-155 ~]$ ./lightfm_optimize_prob.py -threads 1
2017-01-25 16:45:17,436 INFO *main*: Training model...
Epoch 0
2017-01-25 16:52:12,090 INFO *main*: Finished training model

Takes 6min, 55sec.

Now let's look at a non-diverging, gradient doesn't go to zero, param
combo. One thread, no openmp:

$ ./lightfm_optimize_prob.py
/Users/escott/projects/lightfm/lightfm/_lightfm_fast.py:9: UserWarning:
LightFM was compiled without OpenMP support. Only a single thread will be
used.
warnings.warn('LightFM was compiled without OpenMP support. '
2017-01-25 10:50:20,861 INFO *main*: Training model...
...
Epoch 19
2017-01-25 11:01:26,286 INFO *main*: Finished training model

Takes 11min 6sec. This used to take 10min, 30sec before #152
<https://github.com/lyst/lightfm/pull/152>.

Now 16 threads and openmp on AWS:

[hadoop@ip-10-13-181-155 ~]$ ./lightfm_optimize_prob.py
2017-01-25 16:59:48,265 INFO *main*: Training model...
...
Epoch 19
2017-01-25 17:03:57,042 INFO *main*: Finished training model

Takes 4min 9sec. This isn't significantly slower than it used to be, that I
can tell.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/lyst/lightfm/issues/148#issuecomment-275178621>, or mute
the thread

**Comment by scottee** [comment]
One other question then about #152...  Are the results I have from param optimization on version 1.11 still valid for master+PR152?

**Comment by maciejkula** [comment]
I fixed an embarrassing bug with the regularisation, so those results may
unfortunately need re-computing. I don't believe the change will be large,
but it may be worth checking.

**Comment by maciejkula** [comment]
I think this is fixed in 1.12 --- thanks for your help!
